src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Private/JavaScriptMarshalImpl.ToJs.cs

```diff
@@ -0,0 +1,105 @@
+                arg.Int64Value = Unsafe.As<T, long>(ref value);
+                //log.Call(null, $"DynamicMarshalToJs {arg} at {(int)Unsafe.AsPointer(ref arg)}");
+                return;
+            }
+            if (value == null)
+            {
+                arg.Type = IntPtr.Zero;
+                return;
+            }
+
+            // below this is slow path, because of the lookup and also because of boxing.
+            // if you need faster execution, use code generator
+            DynamicMarshalToJsSlow(argType, value, ref arg);
+        }
+
+        private static unsafe void DynamicMarshalToJsSlow(Type argType, object value, ref JavaScriptMarshallerArg arg)
+        {
+            if (marshallers.TryGetValue(argType, out var m1))
+            {
+                // this will match on exact type, including System.Object
+                m1.ToJavaScript(value, ref arg);
+                return;
+            }
+            // this is to support upcasting, for example JSObject implements IJSObject
+            foreach (var m2 in marshallersSequence)
+            {
+                if (m2.ArgumentType.IsAssignableFrom(argType))
+                {
+                    m2.ToJavaScript(value, ref arg);
+                    return;
+                }
+            }
+
+            // fallback System.Object -> gcHandle
+            MarshalToJsObject(value, ref arg);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void InitVoidArgument(ref JavaScriptMarshallerArg arg)
+        {
+#if DEBUG
+            // check alignment
+            Debug.Assert((int)Unsafe.AsPointer(ref arg) % 8 == 0);
+#endif
+            arg.Type = IntPtr.Zero;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void InitArgument<TResult>(ref object root, ref JavaScriptMarshallerArg arg, byte* extra, ref JavaScriptMarshallerArgSignature sig)
+        {
+#if DEBUG
+            // check alignment
+            Debug.Assert((int)Unsafe.AsPointer(ref arg) % 8 == 0);
+#endif
+            arg.Type = sig.Type;
+            if (sig.BufferOffset != -1)
+            {
+                arg.ExtraBufferPtr = (IntPtr)(extra + sig.BufferOffset);
+            }
+            if (sig.UseRoot != 0)
+            {
+                arg.RootRef = (IntPtr)Unsafe.AsPointer(ref root);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void InitExceptionArgument(ref string messageRoot, ref JavaScriptMarshallerArg arg)
+        {
+#if DEBUG
+            // check alignment
+            Debug.Assert((int)Unsafe.AsPointer(ref arg) % 8 == 0);
+#endif
+            arg.Type = exceptionType;
+            arg.RootRef = (IntPtr)Unsafe.AsPointer(ref messageRoot);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void SetRootRef<T>(ref JavaScriptMarshallerArg arg, ref T value)
```

Comment:
The design of this function is fundamentally unsafe - a ref argument is invalid once the function scope exits, but we're storing a pointer to the argument (via Unsafe.AsPointer) in a place that can outlive the function. If we ensure that we're only passing references to something that will outlive the parent scope it's slightly less unsafe but it's still very easy to screw up. I'm not sure how we could best combat this though... I'll give it some thought